Mathlib/CategoryTheory/Functor/OfSequence.lean

```diff
@@ -0,0 +1,130 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Order.Group.Nat
+import Mathlib.CategoryTheory.Category.Preorder
+import Mathlib.CategoryTheory.EqToHom
+
+/-!
+# Functors from the category of the ordered set `ℕ`
+
+In this file, we provide a constructor `Functor.ofSequence`
+for functors `ℕ ⥤ C` which takes as an input a sequence
+of morphisms `f : X n ⟶ X (n + 1)` for all `n : ℕ`.
+
+We also provide a constructor `NatTrans.ofSequence` for natural
+transformations between functors `ℕ ⥤ C` which allows to check
+the naturality condition only for morphisms `n ⟶ n + 1`.
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+variable {C : Type*} [Category C]
+
+namespace Functor
+
+variable {X : ℕ → C} (f : ∀ n, X n ⟶ X (n + 1))
+
+namespace OfSequence
+
+lemma congr_f (i j : ℕ) (h : i = j) :
+    f i = eqToHom (by rw [h]) ≫ f j ≫ eqToHom (by rw [h]) := by
+  subst h
+  simp
+
+/-- The morphism `X i ⟶ X (i + k)` obtained by composing `k` morphisms of
+the form `X n ⟶ X (n + 1)`. -/
+@[simp]
+def map' : ∀ (i k : ℕ), X i ⟶ X (i + k)
+  | _, 0 => 𝟙 _
+  | i, (k+1) => map' i k ≫ f (i + k)
+
+lemma comp_map' (i k₁ k₂ : ℕ) :
+    map' f i k₁ ≫ map' f (i + k₁) k₂ =
+      map' f i (k₁ + k₂) ≫ eqToHom (by rw [add_assoc]) := by
+  revert i k₁
+  induction k₂ with
+  | zero => simp
+  | succ k₂ hk₂ =>
+      intro i k₁
+      simp [reassoc_of% (hk₂ i k₁), congr_f f _ _ (add_assoc i k₁ k₂)]
+
+/-- The morphism `X i ⟶ X j` obtained by composing morphisms of
+the form `X n ⟶ X (n + 1)` when `i ≤ j`. -/
+def map (i j : ℕ) (hij : i ≤ j) : X i ⟶ X j :=
+  map' f i (j - i) ≫ eqToHom (by congr; omega)
```

Comment:
It may be easier to use `Nat.leRecOn` to construct this map.

Reply:
Thanks for the suggestion. I had already tried to make an induction on the `le` type, but it seems it was not possible (because this may eliminate only into props). I could not use `Nat.leRecOn` either, but I have managed to do an induction on the numbers that are involved, which gives good defeq.